privval: race condition in test

```
?   	github.com/tendermint/tendermint/p2p/dummy	[no test files]
==================
WARNING: DATA RACE
Write at 0x00c0001a3551 by goroutine 100:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:589 +0x4c
  testing.(*common).Fatal()
      /usr/local/go/src/testing/testing.go:628 +0x7c
  github.com/tendermint/tendermint/privval.TestListenerConnectDeadlines()
      /go/src/github.com/tendermint/tendermint/privval/socket_test.go:128 +0x268
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162

Previous write at 0x00c0001a3551 by goroutine 94:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:589 +0x4c
  testing.(*common).Fatal()
      /usr/local/go/src/testing/testing.go:628 +0x7c
  github.com/tendermint/tendermint/privval.TestListenerConnectDeadlines.func1()
      /go/src/github.com/tendermint/tendermint/privval/socket_test.go:108 +0xff

Goroutine 100 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:878 +0x659
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:156 +0x332

Goroutine 94 (finished) created at:
  github.com/tendermint/tendermint/privval.TestListenerConnectDeadlines()
      /go/src/github.com/tendermint/tendermint/privval/socket_test.go:103 +0x1e6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:827 +0x162
==================
--- FAIL: TestListenerConnectDeadlines (0.11s)
    socket_test.go:108: read tcp 127.0.0.1:37827->127.0.0.1:38846: i/o timeout
    socket_test.go:128: read tcp 127.0.0.1:38846->127.0.0.1:37827: i/o timeout
    testing.go:771: race detected during execution of test
FAIL
coverage: 87.6% of statements
FAIL	github.com/tendermint/tendermint/privval	1.787s
```

--- a/privval/socket_test.go
+++ b/privval/socket_test.go
@@ -98,36 +98,29 @@ func TestListenerAcceptDeadlines(t *testing.T) {
 
 func TestListenerConnectDeadlines(t *testing.T) {
 	for _, tc := range listenerTestCases(t, time.Second, time.Millisecond) {
-		readyc := make(chan struct{})
-		donec := make(chan struct{})
-		go func(ln net.Listener) {
-			defer close(donec)
-
-			c, err := ln.Accept()
+		go func(dialer Dialer) {
+			_, err := dialer()
 			if err != nil {
-				t.Fatal(err)
+				panic(err)
 			}
-			<-readyc
+		}(tc.dialer)
 
-			time.Sleep(2 * time.Millisecond)
-
-			msg := make([]byte, 200)
-			_, err = c.Read(msg)
-			opErr, ok := err.(*net.OpError)
-			if !ok {
-				t.Fatalf("for %s listener, have %v, want *net.OpError", tc.description, err)
-			}
-
-			if have, want := opErr.Op, "read"; have != want {
-				t.Errorf("for %s listener, have %v, want %v", tc.description, have, want)
-			}
-		}(tc.listener)
-
-		_, err := tc.dialer()
+		c, err := tc.listener.Accept()
 		if err != nil {
 			t.Fatal(err)
 		}
-		close(readyc)
-		<-donec
+
+		time.Sleep(2 * time.Millisecond)
+
+		msg := make([]byte, 200)
+		_, err = c.Read(msg)
+		opErr, ok := err.(*net.OpError)
+		if !ok {
+			t.Fatalf("for %s listener, have %v, want *net.OpError", tc.description, err)
+		}
+
+		if have, want := opErr.Op, "read"; have != want {
+			t.Errorf("for %s listener, have %v, want %v", tc.description, have, want)
+		}
 	}
 }

